Add script to build doxygen

johnnonweiler opened the pull request:
This PR moves the section in .travis.yml responsible for getting the latest version of Doxygen to a separate script, in order to make it easier for Ubuntu users to install it locally.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Add script to build doxygen

Move the section in .travis.yml responsible for getting the latest
version of doxygen to a separate script, in order to make it easier
for Ubuntu users to install it locally.

thk123 commented on .travis.yml:
```diff
@@ -63,10 +63,7 @@ jobs:
-            && wget http://ftp.stack.nl/pub/users/dimitri/doxygen-${DOXYGEN_VERSION}.src.tar.gz -O- | tar -xz --strip-components=1 --directory doxygen \
-            && ( cd doxygen/build && cmake .. ) \
-            && make -j4 -C doxygen/build
+            scripts/build_doxygen.sh
```
You're not calling your script according to its rules shouldn't use be passing in `${DOXYGEN_VERSION}` here?

johnnonweiler replied:
It would have helped if I had typed something along the lines of `git push` after I fixed that problem earlier...  Please could you resolve this conversation before anyone else sees it?

thk123 reviewed (approved):
I like moving stuff out of the travis.yml anyway since it is too large to maintain. But I doubt I'd use it locally since don't like maintaining compiled from source tools - so doesn't really address my comment on the other PR

allredj commented on scripts/build_doxygen.sh:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
```
:no_entry_sign: I think you need `set -euo pipefail` if you want the job to fail when the doxygen build returns non-zero.

allredj replied:
Sorry, I misread. The code is correct as the command is the last one in the script so the return value will be passed on. As discussed offline, it would still be preferable to add the pipefail command and "de-conjunctise" the last command.

allredj reviewed (changes requested):
missing pipefail

Commit:
Refactor build_doxygen.sh to use set -euo pipefail

This refactoring should make it easier to read the script, and less
likely that anyone accidentally breaks the script in future.

johnnonweiler commented:
I've added the requested refactoring as a separate commit, so that the first one is still obviously moving lines from .travis.yml to a separate script.

allredj reviewed (approved):
Thanks for the changes! :+1:

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 603b44aeec488883baf21db706a51fef4ed9b61c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90881837

tautschnig commented:
Needs a rebase to re-trigger the Windows build (not sure why that hasn't happened).

kroening commented:
I would also prefer to use the packaged doxygen -- how about a simple test whether the executable is already there?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 812aa8fb23126e35a4a665d8b4fa4a88ee6b4aba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/90943296
